### ISSUE TITLE
Resque gets stuck on sched_yield when trying to stop with BACKGROUND=yes

### ISSUE THREAD

**Issue by ebeigarts** [resque/resque#578]
Hi,

When trying to stop resque the CPU for the process goes to 100% and strace shows `sched_yield` it had run this way for one week, I can stop it only with `kill -9`.

I can reproduce this only on one server and I don't know what could cause this.

Any ideas?

``` bash
$ PIDFILE=tmp/pids/resque2.pid BACKGROUND=yes VVERBOSE=1 RAILS_ENV=production QUEUE=* bundle exec rake resque:work
```

```
ppoll([{fd=11, events=POLLOUT}], 1, NULL, NULL, 8) = 1 ([{fd=11, revents=POLLOUT}])
write(11, "*2\r\n$3\r\ndel\r\n$39\r\nresque:stat:fa"..., 59) = 59
read(11, ":", 1)                        = 1
read(11, "0\r\n", 8192)                 = 3
getrlimit(RLIMIT_STACK, {rlim_cur=32768*1024, rlim_max=RLIM_INFINITY}) = 0
futex(0x6141074, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x6141070, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
....
```

Without BACKGROUND=1 it kill resque just fine.

``` bash
$ VVERBOSE=1 RAILS_ENV=production QUEUE=* bundle exec rake resque:work
```

```
ppoll([{fd=11, events=POLLOUT}], 1, NULL, NULL, 8) = 1 ([{fd=11, revents=POLLOUT}])
write(11, "*2\r\n$3\r\ndel\r\n$39\r\nresque:stat:fa"..., 59) = 59
read(11, ":", 1)                        = 1
read(11, "0\r\n", 8192)                 = 3
getrlimit(RLIMIT_STACK, {rlim_cur=32768*1024, rlim_max=RLIM_INFINITY}) = 0
futex(0x62526c4, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x62526c0, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x6252728, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x21f2f64, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x21f2f60, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
futex(0x21f2f30, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x21f2f98, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x21f2f9c, FUTEX_WAIT_PRIVATE, 1, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0x21f2f30, FUTEX_WAKE_PRIVATE, 1) = 0
rt_sigaction(SIGINT, {SIG_IGN, [], SA_RESTORER|SA_SIGINFO, 0x7f4d9207f4a0}, {0x4afb50, [], SA_RESTORER|SA_SIGINFO, 0x7f4d9207f4a0}, 8) = 0
rt_sigaction(SIGINT, {SIG_DFL, [], SA_RESTORER|SA_SIGINFO, 0x7f4d9207f4a0}, {SIG_IGN, [], SA_RESTORER|SA_SIGINFO, 0x7f4d9207f4a0}, 8) = 0
close(11)                               = 0
close(9)                                = 0
close(8)                                = 0
write(4, "!", 1)                        = 1
munmap(0x7f4d9239d000, 1052672)         = 0
exit_group(0)                           = ?
Process 8355 detached
```

Machine:
- CentOS 6.2 (2.6.32-220.el6.x86_64)
- ruby 1.9.3p125
- resque-1.20.0

**Comment by drydevelopment** [comment]
I know this is a bit late to the conversation, but the following `capistrano-resque` deploy configuration is working for me. 

```
task :disable_newrelic do
  set :bundle_cmd, "NEWRELIC_ENABLE=false #{bundle_cmd}"
end

before "resque:start", "disable_newrelic"
before "resque:scheduler:start", "disable_newrelic"
```

**Comment by xiaoshengaimm** [comment]
hi all~~
my project do not use newrelic_rpm but has the same problem.

`
[uae@hadoop102 log]$ strace -p 6329
Process 6329 attached - interrupt to quit
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
`

my Gemfile lock is 

`
GEM
  remote: https://rubygems.org/
  specs:
    actionmailer (4.1.2)
      actionpack (= 4.1.2)
      actionview (= 4.1.2)
      mail (~> 2.5.4)
    actionpack (4.1.2)
      actionview (= 4.1.2)
      activesupport (= 4.1.2)
      rack (~> 1.5.2)
      rack-test (~> 0.6.2)
    actionpack-action_caching (1.1.1)
      actionpack (>= 4.0.0, < 5.0)
    actionpack-page_caching (1.0.2)
      actionpack (>= 4.0.0, < 5)
    actionview (4.1.2)
      activesupport (= 4.1.2)
      builder (~> 3.1)
      erubis (~> 2.7.0)
    activemodel (4.1.2)
      activesupport (= 4.1.2)
      builder (~> 3.1)
    activerecord (4.1.2)
      activemodel (= 4.1.2)
      activesupport (= 4.1.2)
      arel (~> 5.0.0)
    activerecord-deprecated_finders (1.0.4)
    activesupport (4.1.2)
      i18n (~> 0.6, >= 0.6.9)
      json (~> 1.7, >= 1.7.7)
      minitest (~> 5.1)
      thread_safe (~> 0.1)
      tzinfo (~> 1.1)
    arel (5.0.1.20140414130214)
    bcrypt (3.1.10)
    bcrypt-ruby (3.1.5)
      bcrypt (>= 3.1.3)
    builder (3.2.2)
    coffee-rails (4.1.0)
      coffee-script (>= 2.2.0)
      railties (>= 4.0.0, < 5.0)
    coffee-script (2.4.1)
      coffee-script-source
      execjs
    coffee-script-source (1.9.1.1)
    daemons (1.2.2)
        devise (3.0.1)
      bcrypt-ruby (~> 3.0)
      orm_adapter (~> 0.1)
      railties (>= 3.2.6, < 5)
      warden (~> 1.2.3)
    devise-encryptable (0.1.2)
      devise (>= 2.1.0)
    erubis (2.7.0)
    eventmachine (1.0.3)
    execjs (2.5.2)
    fast-aes (0.1.2)
    i18n (0.7.0)
    jquery-rails (3.1.2)
      railties (>= 3.0, < 5.0)
      thor (>= 0.14, < 2.0)
    json (1.8.2)
    json_pure (1.8.2)
    mail (2.5.4)
      mime-types (~> 1.16)
      treetop (~> 1.4.8)
    mime-types (1.25.1)
    minitest (5.6.1)
    mysql2 (0.3.18)
    nats (0.4.26)
      daemons (>= 1.1.5)
      eventmachine (>= 0.12.10)
      json_pure (>= 1.7.3)
      thin (>= 1.4.1)
    orm_adapter (0.5.0)
    polyglot (0.3.5)
    protected_attributes (1.0.9)
      activemodel (>= 4.0.1, < 5.0)
    puma (2.7.1)
      rack (>= 1.1, < 2.0)
    rack (1.5.3)
    rack-test (0.6.3)
      rack (>= 1.0)
    rails (4.1.2)
      actionmailer (= 4.1.2)
      actionpack (= 4.1.2)
      actionview (= 4.1.2)
      activemodel (= 4.1.2)
      activerecord (= 4.1.2)
      activesupport (= 4.1.2)
      bundler (>= 1.3.0, < 2.0)
      railties (= 4.1.2)
      sprockets-rails (~> 2.0)
    rails-observers (0.1.2)
      activemodel (~> 4.0)
    railties (4.1.2)
      actionpack (= 4.1.2)
      activesupport (= 4.1.2)
      rake (>= 0.8.7)
      thor (>= 0.18.1, < 2.0)
    rake (10.4.2)
    sass (3.4.14)
    sass-rails (5.0.3)
      railties (>= 4.0.0, < 5.0)
      sass (~> 3.1)
      sprockets (>= 2.8, < 4.0)
      sprockets-rails (>= 2.0, < 4.0)
      tilt (~> 1.1)
    six (0.2.0)
    sprockets (3.1.0)
      rack (~> 1.0)
    sprockets-rails (2.3.1)
      actionpack (>= 3.0)
      activesupport (>= 3.0)
      sprockets (>= 2.8, < 4.0)
    thin (1.6.3)
      daemons (~> 1.0, >= 1.0.9)
      eventmachine (~> 1.0)
      rack (~> 1.0)
    thor (0.19.1)
    thread_safe (0.3.5)
    tilt (1.4.1)
    treetop (1.4.15)
      polyglot
      polyglot (>= 0.3.1)
    tzinfo (1.2.2)
      thread_safe (~> 0.1)
    uglifier (2.7.1)
      execjs (>= 0.3.0)
      json (>= 1.8.0)
    warden (1.2.3)
      rack (>= 1.0)
    yajl-ruby (1.2.1)

PLATFORMS
  ruby

DEPENDENCIES
  actionpack-action_caching
  actionpack-page_caching
  activerecord-deprecated_finders
  coffee-rails (>= 3.2.1)
  devise (= 3.0.1)
  devise-encryptable (= 0.1.2)
  eventmachine (= 1.0.3)
  fast-aes
  jquery-rails
  json
  mysql2
  nats (= 0.4.26)
  protected_attributes
`

**Comment by benweint** [comment]
@xiaoshengaimm as discussed above, the most likely cause of this symptom (with New Relic or not) is a [bug in Ruby](https://bugs.ruby-lang.org/issues/7720) related to the behavior of `Process.daemon`, which was fixed in Ruby 2.0.0. This bug can be triggered by any code that spawns a background thread before Resque calls `Process.daemon`.

My recommendation would be to upgrade to Ruby 2.0.0 or later.

**Comment by xiaoshengaimm** [comment]
@benweint ok~ thx
